src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,166 @@
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAddedElement(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> succ = head;
+            int i = 0;
+            while (index != i) {
+                succ = succ.next;
+                i++;
+            }
+            Node<T> pred = succ.prev;
+            Node<T> newNode = new Node<>(pred, value, succ);
+            succ.prev = newNode;
+            if (pred == null) {
+                head = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currNode;
+        if (index < (size >> 1)) {
+            currNode = head;
+            for (int i = 0; i < index; i++) {
+                currNode = currNode.next;
+            }
+        } else {
+            currNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currNode = currNode.prev;
+            }
+        }
+        return currNode.item;
     }
 
     @Override
     public T set(T value, int index) {
+        checkIndex(index);
+        Node<T> currNode = head;
+        for (int i = 0; i <= index; i++) {
+            if (index == i) {
+                T oldItem = currNode.item;
+                currNode.item = value;
+                return oldItem;
+            }
+            currNode = currNode.next;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndex(index);
+        Node<T> nodeToDelete = head;
+        for (int i = 0; i <= index; i++) {
+            if (index == i) {
+                T element = nodeToDelete.item;
+                Node<T> next = nodeToDelete.next;
+                Node<T> prev = nodeToDelete.prev;
+                if (prev == null) {
+                    head = next;
+                } else {
+                    prev.next = nodeToDelete.next;
+                    nodeToDelete.next = null;
+                }
+                if (next == null) {
+                    tail = prev;
+                } else {
+                    next.prev = nodeToDelete.prev;
+                    nodeToDelete.prev = null;
+                }
+                nodeToDelete.item = null;
+                size--;
+                return element;
+            }
+            nodeToDelete = nodeToDelete.next;
+        }
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        int indexToRemove = 0;
+        Node<T> currItem = head;
+        while (currItem != null) {
+            if (Objects.equals(currItem.item, object)) {
+                remove(indexToRemove);
```

Comment:
Calling `remove(int index)` within `remove(T object)` is not optimal as it traverses the list twice. Consider directly manipulating the nodes to remove the element for better performance.